contracts/utils/TokenVestingStaking.sol

```diff
@@ -0,0 +1,260 @@
+    //                                                          //
+    //////////////////////////////////////////////////////////////
+
+    /**
+     * @notice Adds a BLS public key to the list of service nodes. Requires a proof of possession BLS signature to prove user controls the public key being added
+     * @param blsPubkey - 64 bytes of the bls public key
+     * @param blsSignature - 128 byte signature
+     * @param serviceNodeParams - Service node public key, signature proving ownership of public key and fee that operator is charging
+     */
+    function addBLSPublicKey(BN256G1.G1Point calldata blsPubkey, IServiceNodeRewards.BLSSignatureParams calldata blsSignature, IServiceNodeRewards.ServiceNodeParams calldata serviceNodeParams) external onlyBeneficiary notRevoked afterStart {
+        uint256 stakingRequirement = stakingRewardsContract.stakingRequirement();
+        uint64 serviceNodeID = stakingRewardsContract.nextServiceNodeID();
+        SENT.approve(address(stakingRewardsContract), stakingRequirement);
+        serviceNodes.push(ServiceNode(serviceNodeID, stakingRequirement));
+        IServiceNodeRewards.Contributor[] memory contributors = new IServiceNodeRewards.Contributor[](1);
+        contributors[0] = IServiceNodeRewards.Contributor(address(this), stakingRequirement);
+        stakingRewardsContract.addBLSPublicKey(blsPubkey, blsSignature, serviceNodeParams, contributors);
+    }
+
+    /**
+     * @notice Starts the process for removing a Service Node
+     * @param serviceNodeID the identifier of the node which is being removed from network
+     */
+    function initiateRemoveBLSPublicKey(uint64 serviceNodeID) external onlyBeneficiary notRevoked afterStart {
+        stakingRewardsContract.initiateRemoveBLSPublicKey(serviceNodeID);
+    }
+
+    /**
+     * @notice Gets rewards from staking and transfers to beneficiary
+     */
+    function claimRewards() external onlyBeneficiary notRevoked afterStart {
+        uint256 unstaked;
+
+        uint256 length = serviceNodes.length;
+
+        for (uint256 i = 1; i < length + 1; i++) {
+            IServiceNodeRewards.ServiceNode memory serviceNode = stakingRewardsContract.serviceNodes(serviceNodes[i - 1].serviceNodeID);
+            if (serviceNode.deposit == 0) {
+                unstaked += serviceNodes[i - 1].deposit;
+                
+                // Remove service node from the array by swapping it with the last element and then popping the array
+                serviceNodes[i - 1] = serviceNodes[length - 1];
+                serviceNodes.pop();
+                
+                // Adjust loop variables since we modified the array
+                i--;
+                length--;
+            }
+        }
+
+        uint256 balanceBeforeClaiming = SENT.balanceOf(address(this));
+        stakingRewardsContract.claimRewards();
+        uint256 balanceAfterClaiming = SENT.balanceOf(address(this)) - unstaked;
```

Comment:
After some thinking, I think `unstaked` here is not correct. I couldn't figure out a stiutation where unstaked nodes will _not_ be reflected in `SENT.balanceOf(address(this))` already by the time we issue these 2 calls.

`claimRewards` doesn't have any side effects in regards to the deposit balances of service nodes that were liquidated or removed.